bin/upgrade.py

```diff
@@ -116,14 +119,50 @@ def rename_testcase(old_base: Path, new_dir: Path) -> None:
     if bad_dir.is_dir() and not any(bad_dir.iterdir()):
         bad_dir.rmdir()
 
+    # Move .hint and .desc files to the Test Case Configuration .yaml file
+
+    test_case_yamls = defaultdict[Path, CommentedMap](CommentedMap)
+    for f in (problem_path / "data").rglob("*.yaml"):
+        if f.with_suffix(".in").exists():  # Prevent reading test_group.yaml, which has no *.in file
+            test_case_yamls[f] = read_yaml(f)
 
-def upgrade_testdata_yaml(problem_path: Path, bar: ProgressBar) -> None:
+    for f in (problem_path / "data").rglob("*.desc"):
+        test_case_yaml = test_case_yamls[f.with_suffix(".yaml")]
+        if "description" in test_case_yaml:
+            bar.warn(f"can't move '{f}' to '*.yaml', it already contains the key 'description'")
+        else:
+            bar.log(f"moving '{f}' to 'description' key in '*.yaml'")
+            test_case_yaml["description"] = f.read_text()
+            write_yaml(test_case_yaml, f.with_suffix(".yaml"))
+            f.unlink()
+
+    for f in (problem_path / "data").rglob("*.hint"):
+        test_case_yaml = test_case_yamls[f.with_suffix(".yaml")]
+        if "hint" in test_case_yaml:
+            bar.warn(f"can't move '{f}' to '*.yaml', it already contains the key 'hint'")
+        else:
+            bar.log(f"moving '{f}' to 'hint' key in '*.yaml'")
+            test_case_yaml["hint"] = f.read_text()
+            write_yaml(test_case_yaml, f.with_suffix(".yaml"))
+            f.unlink()
+
+
+def rename_testdata_to_test_group_yaml(problem_path: Path, bar: ProgressBar) -> None:
+    for f in (problem_path / "data").rglob("testdata.yaml"):
+        new_name = f.with_name("test_group.yaml")
```

Comment:
error if `new_name` already exists. (For a proper problem in legacy format this won't happen but problems might be a weird mix of legacy/draft...)